### PR TITLE
Allow restore of empty partitioned table from snapshot

### PR DESCRIPTION
**Pull request by bogensberger**
PR to ES-Patch: https://github.com/crate/elasticsearch/pull/113

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java [hunk 1]
```diff
@@ -151,10 +152,13 @@ public void onFailure(Exception e) {
         IndicesOptions indicesOptions = IndicesOptions.fromOptions(ignoreUnavailable, true, true, false, IndicesOptions.lenientExpandOpen());
         FutureActionListener<RestoreSnapshotResponse, Long> listener = new FutureActionListener<>(r -> 1L);
         resolveIndexNames(analysis.restoreTables(), ignoreUnavailable, transportActionProvider.transportGetSnapshotsAction(), analysis.repositoryName())
-            .whenComplete((List<String> indexNames, Throwable t) -> {
+            .whenComplete((ResolveIndicesAndTemplatesContext resolvedIndicesAndSnapshots, Throwable t) -> {
                 if (t == null) {
+                    List<String> indexNames = ImmutableList.copyOf(resolvedIndicesAndSnapshots.resolvedIndices);
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: can we avoid copying the collection/set here? at least, create an string array, otherwise all would be copied again at RestoreSnapshoptRequest

File: CHANGES.txt [hunk 2]
```diff
@@ -110,7 +110,7 @@ Breaking Changes
 
 Changes
 =======
-
+
```
Comment (seut) [inline CHANGES.txt]: pls remove space

File: CHANGES.txt [hunk 3]
```diff
@@ -145,6 +145,10 @@ Changes
 Fixes
 =====
 
+ - Fix: Restoring empty partitioned table wasn't possible
+
+    - Updated elasticsearch upstream to allow restore of partition templates
```
Comment (seut) [inline CHANGES.txt]: this is unrelated for crate users, pls remove

File: CHANGES.txt [hunk 4]
```diff
@@ -145,6 +145,10 @@ Changes
 Fixes
 =====
 
+ - Fix: Restoring empty partitioned table wasn't possible
```
Comment (seut) [inline CHANGES.txt]: Restoring **a**..  or Restoring... table**s**

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java [hunk 5]
```diff
@@ -169,77 +173,96 @@ public void onFailure(Exception e) {
...
-        Collection<String> resolvedIndices = new HashSet<>();
+    static CompletableFuture<ResolveIndicesAndTemplatesContext> resolveIndexNames(List<RestoreSnapshotAnalyzedStatement.RestoreTableInfo> restoreTables, boolean ignoreUnavailable,
+                                                                                  TransportGetSnapshotsAction getSnapshotsAction, String repositoryName) {
+        ResolveIndicesAndTemplatesContext resolveIndicesAndTemplatesContext = new ResolveIndicesAndTemplatesContext();
         List<RestoreSnapshotAnalyzedStatement.RestoreTableInfo> toResolveFromSnapshot = new ArrayList<>();
         for (RestoreSnapshotAnalyzedStatement.RestoreTableInfo table : restoreTables) {
             if (table.hasPartitionInfo()) {
-                resolvedIndices.add(table.partitionName().asIndexName());
+                resolveIndicesAndTemplatesContext.addIndex(table.partitionName().asIndexName());
+                resolveIndicesAndTemplatesContext.addTemplate(PartitionName.templateName(table.tableIdent().schema(), table.tableIdent().name()));
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: think it would make sense to add the templateName to RestoreTableInfo to avoid repeating..

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java [hunk 6]
```diff
@@ -169,77 +173,96 @@ public void onFailure(Exception e) {
...
             new GetSnapshotsRequest(repositoryName),
-            new ResolveFromSnapshotActionListener(f, toResolveFromSnapshot, resolvedIndices)
+            new ResolveFromSnapshotActionListener(f, toResolveFromSnapshot, resolveIndicesAndTemplatesContext)
         );
         return f;
     }
 
+    static class ResolveIndicesAndTemplatesContext {
+
+        Collection<String> resolvedIndices = new HashSet<>();
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: can be final and private

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java [hunk 7]
```diff
@@ -169,77 +173,96 @@ public void onFailure(Exception e) {
...
+        Collection<String> resolvedTemplates = new HashSet<>();
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: also private final

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java [hunk 8]
```diff
@@ -169,77 +173,96 @@ public void onFailure(Exception e) {
...
+                resolveTableFromSnapshot(table.tableIdent(), snapshots, resolveIndicesAndTemplatesContext);
             }
-            returnFuture.complete(ImmutableList.copyOf(resolvedIndices));
+            returnFuture.complete(resolveIndicesAndTemplatesContext);
         }
 
         @VisibleForTesting
-        public static String resolveIndexNameFromSnapshot(TableIdent tableIdent, List<SnapshotInfo> snapshots) throws TableUnknownException{
-            String indexName = tableIdent.indexName();
+        public static void resolveTableFromSnapshot(TableIdent tableIdent, List<SnapshotInfo> snapshots,
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: pls put snapshots arg in new line

File: CHANGES.txt (hunk 3 above)
Comment (mfussenegger) [inline CHANGES.txt]: I'd remove it, it's not relevant for a user which parts of the code-base changed in order to fix the issue

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java [hunk 9]
```diff
@@ -169,77 +173,96 @@ public void onFailure(Exception e) {
...
                         // to match all partitions if a partitioned table was meant
-                        return PartitionName.templateName(tableIdent.schema(), tableIdent.name()) + "*";
+                        ctx.addIndex(templateName + "*");
+                        ctx.addTemplate(templateName);
+                        return;
                     }
                 }
             }
-            throw new TableUnknownException(tableIdent);
+            ctx.addTemplate(PartitionName.templateName(tableIdent.schema(), tableIdent.name()));
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: pls avoid creating the templateName twice

File: CHANGES.txt (hunk 3 above)
Comment (bogensberger) [inline CHANGES.txt]: you are right, updated.

File: sql/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java [hunk 10]
```diff
@@ -374,4 +374,52 @@ public void testRestoreOnlyOnePartitionedTable() throws Exception {
         execute("select table_schema || '.' || table_name from information_schema.tables where table_schema='doc'");
         assertThat(TestingHelpers.printedTable(response.rows()), is("doc.my_parted_1\n"));
     }
+
+    @Test
+    public void testRestoreEmptyPartitionedTableUsingALL() throws Exception {
+        execute("create table employees(section integer, name string) partitioned by (section)");
+        ensureYellow();
+        execute("refresh table employees");
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java]: refresh is not needed here, pls remove

File: sql/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java [hunk 11]
```diff
@@ -374,4 +374,52 @@ public void testRestoreOnlyOnePartitionedTable() throws Exception {
...
+
+        execute("select table_schema || '.' || table_name from information_schema.tables where table_schema='doc'");
+        assertThat(TestingHelpers.printedTable(response.rows()), is("doc.employees\n"));
+    }
+
+    @Test
+    public void testRestoreEmptyPartitionedTable() throws Exception {
+        execute("create table employees(section integer, name string) partitioned by (section)");
+        ensureYellow();
+        execute("refresh table employees");
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java]: same here

File: sql/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java [hunk 12]
```diff
@@ -374,4 +374,52 @@ public void testRestoreOnlyOnePartitionedTable() throws Exception {
...
+
+        execute("select table_schema || '.' || table_name from information_schema.tables where table_schema='doc'");
+        assertThat(TestingHelpers.printedTable(response.rows()), is("doc.employees\n"));
+    }
+
+    @Test
+    public void testResolveAllFromSnapshotWithoutAnyTable() throws Exception {
+        execute("CREATE SNAPSHOT " + snapshotName() + " ALL WITH (wait_for_completion=true)");
+        ensureYellow();
+        execute("RESTORE SNAPSHOT " + snapshotName() + " ALL with (wait_for_completion=true)");
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java]: what is this test doing? it's not asserting anything...

File: sql/src/test/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcherTest.java [hunk 13]
```diff
@@ -44,35 +48,57 @@
...
+        SnapshotRestoreDDLDispatcher.ResolveIndicesAndTemplatesContext ctx = new SnapshotRestoreDDLDispatcher.ResolveIndicesAndTemplatesContext();
+        SnapshotRestoreDDLDispatcher.ResolveFromSnapshotActionListener.resolveTableFromSnapshot(
+            new TableIdent(null, "restoreme"),
+            Collections.singletonList(
+                new SnapshotInfo(new SnapshotId("snapshot01", UUID.randomUUID().toString()), ImmutableList.of(), 0L)
+            ),
+            ctx
+        );
+        assertThat(ctx.resolvedIndices.size(), is(0));
+        assertThat(ctx.resolvedTemplates, contains(PartitionName.templateName(null, "restoreme")));
```
Comment (seut) [inline sql/src/test/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcherTest.java]: pls add a comment why a template should be restored in that case, nothing beside the test name is partition related...

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java (hunk 6 above)
Comment (bogensberger) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: as i use it in the tests i need it package private. Made it final and added the @VisibleForTesting annotation to the class.

**Comment by bogensberger** [comment]
@seut addressed your comments and also updated: https://github.com/crate/elasticsearch/pull/113

File: sql/src/main/java/io/crate/analyze/RestoreSnapshotAnalyzedStatement.java [hunk 14]
```diff
@@ -102,5 +102,9 @@ public PartitionName partitionName() {
         public boolean hasPartitionInfo() {
             return partitionName != null;
         }
+
+        public String createPartitionsTemplateName() {
+            return PartitionName.templateName(tableIdent.schema(), tableIdent().name());
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/RestoreSnapshotAnalyzedStatement.java]: can we save this instead of creating it always?

File: sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java [hunk 15]
```diff
@@ -205,10 +205,11 @@ public void onFailure(Exception e) {
         return f;
     }
 
+    @VisibleForTesting
     static class ResolveIndicesAndTemplatesContext {
 
-        Collection<String> resolvedIndices = new HashSet<>();
-        Collection<String> resolvedTemplates = new HashSet<>();
+        final Collection<String> resolvedIndices = new HashSet<>();
```
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java]: why aren't they private? can't we make them private? otherwise the add methods doesn't make so much sense or?
